Northstar.CustomServers/mod/scripts/vscripts/_custom_codecallbacks.gnut

```diff
@@ -1,44 +1,56 @@
 untyped
 
-globalize_all_functions
+global function AddCallback_OnReceivedSayTextMessage
+global function NSSetupChathooksServer
 
 global struct ClServer_MessageStruct {
 	string message
 	entity player
-	int channelId
+	bool isTeam
 	bool shouldBlock
 }
 
 struct {
 	array< ClServer_MessageStruct functionref( ClServer_MessageStruct ) > OnReceivedSayTextMessageCallbacks
 } NsCustomCallbacks
 
-void function CServerGameDLL_ProcessMessageStartThread()
+void function CServerGameDLL_ProcessMessageStartThread(int playerIndex, string message, bool isTeam)
 {
-	thread CServerGameDLL_OnReceivedSayTextMessageCallback()
+	thread CServerGameDLL_OnReceivedSayTextMessageCallback(playerIndex, message, isTeam)
 }
 
-void function CServerGameDLL_OnReceivedSayTextMessageCallback()
+void function CServerGameDLL_OnReceivedSayTextMessageCallback(int playerIndex, string message, bool isTeam)
 {
+	entity player = GetPlayerByIndex(playerIndex)
+	if (player == null) {
+		print("Ignored chat message from invalid player index " + playerIndex + ": " + message)
+		return
+	}
+
 	ClServer_MessageStruct localMessage
-	localMessage.message =  NSChatGetCurrentMessage()
-	localMessage.player = GetPlayerByIndex(NSChatGetCurrentPlayer())
-	localMessage.channelId = NSChatGetCurrentChannel()
+	localMessage.message = message
+	localMessage.player = player
+	localMessage.isTeam = isTeam
 	localMessage.shouldBlock = false
-
 	foreach ( callbackFunc in NsCustomCallbacks.OnReceivedSayTextMessageCallbacks )
 	{
 		ClServer_MessageStruct returnStruct = callbackFunc(localMessage)
 		localMessage.message = returnStruct.message
-		localMessage.player = returnStruct.player
-		localMessage.channelId = returnStruct.channelId
-		localMessage.shouldBlock = localMessage.shouldBlock || returnStruct.shouldBlock
+		localMessage.isTeam = returnStruct.isTeam
+		if (localMessage.shouldBlock)
+		{
+			return
+		}
 	}
 
-	NSSetMessage(localMessage.message, localMessage.player.GetPlayerIndex(), localMessage.channelId, localMessage.shouldBlock)
+	NSSendMessage(player.GetPlayerIndex(), localMessage.message, localMessage.isTeam)
```

Comment:
IMO all mods subscribed to the message handler should see the message, even if an earlier mod decides it wants to block it. You could simply `|` `localMessage.shouldBlock` like in my original implementation as discussed on discord ^^